File: sdks/go/entities.go
package atap

import "context"

// EntityAPI provides entity registration, retrieval, deletion, and key rotation.
type EntityAPI struct {
	client *Client
}

// RegisterOptions holds optional parameters for entity registration.
type RegisterOptions struct {
	Name         string
	PublicKey    string
	PrincipalDID string
}

// Register creates a new entity.
//
// entityType is one of "agent", "machine", "human", "org".
// For agent/machine, the response includes client_secret (returned once).
// If no public key is provided, the server generates a keypair and returns private_key (once).
func (e *EntityAPI) Register(ctx context.Context, entityType string, opts *RegisterOptions) (*Entity, error) {
	body := map[string]interface{}{
		"type": entityType,
	}
	if opts != nil {
		if opts.Name != "" {
			body["name"] = opts.Name
		}
		if opts.PublicKey != "" {
			body["public_key"] = opts.PublicKey
		}
		if opts.PrincipalDID != "" {
			body["principal_did"] = opts.PrincipalDID
		}
	}

	data, err := e.client.http.Request(ctx, "POST", "/v1/entities", &RequestOptions{JSONBody: body})
	if err != nil {
		return nil, err
	}
	return parseEntity(data), nil
}

// Get retrieves public entity info by ID.
func (e *EntityAPI) Get(ctx context.Context, entityID string) (*Entity, error) {
	data, err := e.client.http.Request(ctx, "GET", "/v1/entities/"+entityID, nil)
	if err != nil {
		return nil, err
	}
	return parseEntity(data), nil
}

// Delete removes an entity (crypto-shred). Requires atap:manage scope.
func (e *EntityAPI) Delete(ctx context.Context, entityID string) error {
	_, err := e.client.authedRequest(ctx, "DELETE", "/v1/entities/"+entityID, nil)
	return err
}

// RotateKey rotates an entity's Ed25519 public key. Requires atap:manage scope.
func (e *EntityAPI) RotateKey(ctx context.Context, entityID, publicKey string) (*KeyVersion, error) {
	body := map[string]interface{}{
		"public_key": publicKey,
	}
	data, err := e.client.authedRequest(ctx, "POST", "/v1/entities/"+entityID+"/keys/rotate", &RequestOptions{JSONBody: body})
	if err != nil {
		return nil, err
	}
	return &KeyVersion{
		ID:         getString(data, "id"),
		EntityID:   getString(data, "entity_id"),
		KeyIndex:   getInt(data, "key_index", 0),
		ValidFrom:  getString(data, "valid_from"),
		ValidUntil: getString(data, "valid_until"),
		CreatedAt:  getString(data, "created_at"),
	}, nil
}

func parseEntity(data map[string]interface{}) *Entity {
	return &Entity{
		ID:           getString(data, "id"),
		Type:         getString(data, "type"),
		DID:          getString(data, "did"),
		PrincipalDID: getString(data, "principal_did"),
		Name:         getString(data, "name"),
		KeyID:        getString(data, "key_id"),
		PublicKey:    getString(data, "public_key"),
		TrustLevel:   getInt(data, "trust_level", 0),
		Registry:     getString(data, "registry"),
		CreatedAt:    getString(data, "created_at"),
		UpdatedAt:    getString(data, "updated_at"),
		ClientSecret: getString(data, "client_secret"),
		PrivateKey:   getString(data, "private_key"),
	}
}

File: sdks/javascript/src/entities.ts
/**
 * Entity management operations.
 */

import type { ATAPClient } from "./client.js";
import type { Entity, KeyVersion } from "./models.js";

/** Entity registration, retrieval, deletion, and key rotation. */
export class EntityAPI {
  private readonly _client: ATAPClient;

  constructor(client: ATAPClient) {
    this._client = client;
  }

  /**
   * Register a new entity.
   *
   * @param entityType - One of "agent", "machine", "human", "org".
   * @param options.name - Optional display name.
   * @param options.publicKey - Optional base64-encoded Ed25519 public key.
   * @param options.principalDid - Optional DID for agent-to-principal binding.
   * @returns Entity with id, did, type, name, keyId. For agent/machine: also includes clientSecret.
   */
  async register(
    entityType: string,
    options: {
      name?: string;
      publicKey?: string;
      principalDid?: string;
    } = {},
  ): Promise<Entity> {
    const body: Record<string, string> = { type: entityType };
    if (options.name) body.name = options.name;
    if (options.publicKey) body.public_key = options.publicKey;
    if (options.principalDid) body.principal_did = options.principalDid;

    const data = await this._client._http.request("POST", "/v1/entities", {
      jsonBody: body,
    });
    return parseEntity(data);
  }

  /** Get public entity info by ID. */
  async get(entityId: string): Promise<Entity> {
    const data = await this._client._http.request(
      "GET",
      `/v1/entities/${entityId}`,
    );
    return parseEntity(data);
  }

  /** Delete an entity (crypto-shred). Requires atap:manage scope. */
  async delete(entityId: string): Promise<void> {
    await this._client._authedRequest("DELETE", `/v1/entities/${entityId}`);
  }

  /**
   * Rotate an entity's Ed25519 public key. Requires atap:manage scope.
   *
   * @param entityId - The entity ID.
   * @param publicKey - Base64-encoded new Ed25519 public key.
   * @returns New KeyVersion with id, keyIndex, validFrom.
   */
  async rotateKey(entityId: string, publicKey: string): Promise<KeyVersion> {
    const data = await this._client._authedRequest(
      "POST",
      `/v1/entities/${entityId}/keys/rotate`,
      { jsonBody: { public_key: publicKey } },
    );
    return {
      id: (data.id as string) || "",
      entityId: (data.entity_id as string) || "",
      keyIndex: (data.key_index as number) || 0,
      validFrom: (data.valid_from as string) || "",
      validUntil: data.valid_until as string | undefined,
      createdAt: (data.created_at as string) || "",
    };
  }
}

export function parseEntity(data: Record<string, unknown>): Entity {
  return {
    id: (data.id as string) || "",
    type: (data.type as string) || "",
    did: (data.did as string) || "",
    principalDid: (data.principal_did as string) || "",
    name: (data.name as string) || "",
    keyId: (data.key_id as string) || "",
    publicKey: (data.public_key as string) || "",
    trustLevel: (data.trust_level as number) || 0,
    registry: (data.registry as string) || "",
    createdAt: (data.created_at as string) || "",
    updatedAt: (data.updated_at as string) || "",
    clientSecret: data.client_secret as string | undefined,
    privateKey: data.private_key as string | undefined,
  };
}
